### PR TITLE
Moving to async/await: lib/Runner.js

### LINKED ISSUE #463: Replace `new Promise` and `.then` with async/await
We no longer support node 8 so there is nothing holding us back from asyncing and awaiting.

---

To do this

1. Search in files for `new Promise` or `.then`.
2. Replace it with an async function.
3. Make a PR.

It is worth keeping the PR's small (maybe only adding a one or two async/awaits at a time) so we can review them quickly and get them merged!

### PR DESCRIPTION
**Pull request by chebro**
This PR attempts to close #463

### CODE REVIEW

**Comment by sravanth-chebrolu** [comment]
Just checked Travis log, I found this:
```
mod.possiblyTests.map is not a function
``` 
Am I doing something wrong?

**Comment by harrysarson** [comment]
@lydell this is a potential downside of the better error handling we have rolled out. We do no longer get stack traces for errors like these.

@sravanth-chebrolu it looks like somewhere you are returning an object containing a `Promise` rather than a `Promise` to an object. Try running `npm check` I think flow should tell you where you are returning a slightly wrong thing.

**Comment by lydell** [comment]
@harrysarson Flow should catch type errors.

**Comment by sravanth-chebrolu** [comment]
Yeah I found the error, it was a stupid mistake, I forgot the `await` keyword. Sorry for wasting your time :(

**Comment by sravanth-chebrolu** [comment]
Flow still throws this error

```
> flow

Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ lib/Runner.js:54:35

Unexpected identifier, expected the token ;

     51│         );
     52│       }
     53│
     54│       const possiblyTests = await Parser.extractExposedPossiblyTests(filePath);
     55│       return { 
     56│         moduleName, 
     57│         possiblyTests 

```

Any help? I couldn't figure this out on my own

**Comment by sravanth-chebrolu** [comment]
Weird.. tests pass for Nodejs v14, but fail for the rest

File: lib/elm-test.js [hunk 1]
```diff
@@ -205,32 +205,29 @@ function runTests(generatedCodeDir /*: string */, testFile /*: string */) {
...
-            dest,
-            args.watch,
-            Compile.isMachineReadableReporter(report)
-          );
-        }
-      );
-    })
-    .catch(function (error) {
+  await Generate.prepareCompiledJsFile(pipeFilename, dest)
+  try {
```
Comment (harrysarson) [inline lib/elm-test.js:216]:
```suggestion
  try {
    await Generate.prepareCompiledJsFile(pipeFilename, dest)
```

The `.catch` in the promise chain catches errors from _all_ previous promises not just the last one.

File: lib/elm-test.js [hunk 2]
```diff
@@ -333,19 +330,22 @@ if (isMake) {
...
-    });
+  (async () => {
+    await Compile.compileSources(
+      testFilePaths,
+      generatedCodeDir,
+      args.verbose,
+      pathToElmBinary,
+      args.report
+    )
+    try {
```
Comment (harrysarson) [inline lib/elm-test.js:341]:
```suggestion
    try {
    await Compile.compileSources(
      testFilePaths,
      generatedCodeDir,
      args.verbose,
      pathToElmBinary,
      args.report
    )
```

File: lib/elm-test.js [hunk 3]
```diff
@@ -360,35 +360,33 @@ if (isMake) {
...
-          parseInt(args.seed),
-          args.report,
-          testFileGlobs,
-          testFilePaths,
-          testModules,
-          generatedSrc,
-          processes
-        );
+    const testModules = await Runner.findTests(testFilePaths, sourceDirs, isPackageProject)
+    try {
```
Comment (harrysarson) [inline lib/elm-test.js:367]:
```suggestion
    try {
      const testModules = await Runner.findTests(testFilePaths, sourceDirs, isPackageProject)
```

File: lib/elm-test.js (hunk 2 above)
Comment (harrysarson) [reply lib/elm-test.js:341]: Indentation will be wrong, but run `npm run format:write` to fix.

**Review by harrysarson — COMMENTED** [review]
CI failures are because you need to put more of the function calls within the try catch

**Review by sravanth-chebrolu — COMMENTED** [review]
Thanks I see the mistakes, I've learnt so much about promises from just this PR..

**Comment by chebro** [comment]
Should we merge #465? I don't understand why the checks are still failing

**Comment by lydell** [comment]
@chebro The checks are failing because ESLint fails to parse `catch` without `(foo)` after it. This can be solved in three ways:

1. Use `catch (_error)` instead.
2. Bump `ecmaVersion` in .eslintrc.js from `2018` to `2019`. I’m not sure if this is safe in general for Node.js 10+ support.
3. Merge #465 first which does not this problem – when I converted to async/await there I did not end up ignoring any errors.

**Comment by chebro** [comment]
> 1. Use `catch (_error)` instead.
Will do this 👍🏽 
> 3. Merge #465 first which does not this problem – when I converted to async/await there I did not end up ignoring any errors.
Should I accept all incoming changes?

**Comment by lydell** [comment]
> Should I accept all incoming changes?

Yes, that’s the plan once #465 is merged into master. (Note: There are one or two `.then` left in my version where I considered that more readable.) We’ll see which PR is merged first and who gets to solve the conflicts :)

**Comment by chebro** [comment]
This is my first merge on github, hope I did it right 🤞

**Comment by lydell** [comment]
@chebro I didn’t mean that you should merge #465 into this branch yet. But maybe things will sort themselves out if we merge #465 into master and then squash merge this one. 🤔

**Comment by chebro** [comment]
I'll just wait until your branch is merged then

**Comment by harrysarson** [comment]
CI is green 👍 marking as blocked as I cannot review due to all the changes pulled in from the other PR?

**Comment by chebro** [comment]
> due to all the changes pulled in from the other PR?

Yes, I misunderstood was @lydell said, my bad.

**Comment by harrysarson** [comment]
I am happy to merge this before #465 if it gets untangled. Equally I am very happy to wait :)

**Comment by harrysarson** [comment]
You are free to go for it @chebro!

**Comment by chebro** [comment]
> You are free to go for it @chebro!

Aye! I'm a little busy this week, will start as soon as I can.

**Comment by harrysarson** [comment]
No worries! Take your time (amongst other things of ci is broken so we can't merge until that is fixed anyway).

**Comment by chebro** [comment]
Hmm, there seem to be some conflicts now, should I pull the master branch into this?

**Comment by harrysarson** [comment]
Yeah! Don't worry about the commit history looking nice because I will squash when I merge

**Comment by lydell** [comment]
Also note that this PR is not needed anymore if #475 is merged. Sorry!

**Comment by chebro** [comment]
> Also note that this PR is not needed anymore if #475 is merged. Sorry!

Alright then, we can close this PR. I'm glad that I got to take away something I previously didn't know about promises from this PR. Good luck :)

**Comment by harrysarson** [comment]
Thanks @chebro, as they say "code is the easy part". Most the valuable work in open source is exploring the design space and working out  what code you actually need, that is to say: PRs that do not get merged are no less valuable to a project as those that do. Till next time ...

**Pull request closed without merging**